packages/server/src/connector/index.ts

```diff
@@ -208,71 +204,102 @@ async function subscribeEvents() {
   })
 
   connector.addSubscription(unsubscribe)
-  return
+}
+
+async function subscribeFinalizedHeaders() {
+  const unsubscribe = await api.rpc.chain.subscribeFinalizedHeads(
+    async header => {
+      let hash = header.hash.toString()
+      let number = header.number.toNumber()
+      let missing = 0
+      let enhancedHeader = await getBlockHeader(number)
+
+      if (!firstSavedBlock.number)
+        firstSavedBlock = { number, hash, timestamp: enhancedHeader.timestamp }
+
+      missing = lastSavedBlock.number
+        ? number - lastSavedBlock.number - 1
+        : number - firstSavedBlock.number
+
+      lastSavedBlock = { number, hash, timestamp: enhancedHeader.timestamp }
+
+      //GET Missing blocks, TODO: this could make hole in the data if missing
+      //server is turned off while getting missing blocks
+
+      console.log('finalized header #:', number, 'with hash:', hash)
+
+      if (headerCache[number]) {
+        if (settings.get().validatorId === enhancedHeader.author)
+          testEquivocation(headerCache[number], enhancedHeader)
+
+        headerCache[number] = null
+        delete headerCache[number]
+      }
+
+      if (missing > 0) {
+        console.log('missing', missing, 'headers')
+        let missingHeaders = await getPreviousHeaders(
+          missing,
+          lastSavedBlock.number
+        )
+
+        await db.bulkSave('Header', missingHeaders)
+      }
+
+      await db.saveHeader(enhancedHeader)
+    }
+  )
+
+  connector.addSubscription(unsubscribe)
 }
 
 async function subscribeHeaders() {
   const unsubscribe = await api.rpc.chain.subscribeNewHeads(async header => {
     let hash = header.hash.toString()
     let number = header.number.toNumber()
-    let missing = 0
     let enhancedHeader = await getBlockHeader(number)
     let finalizedHash = await api.rpc.chain.getFinalizedHead()
-    watcher.ping(enhancedHeader.timestamp, finalizedHash.toString())
-
-    console.log('new header #:', number, 'with hash:', hash)
 
     analyzeExtrinsics(hash)
 
-    if (!firstSavedBlock.number)
-      firstSavedBlock = { number, hash, timestamp: enhancedHeader.timestamp }
-
-    missing = lastSavedBlock.number
-      ? number - lastSavedBlock.number - 1
-      : number - firstSavedBlock.number
-
-    lastSavedBlock = { number, hash, timestamp: enhancedHeader.timestamp }
-
-    //GET Missing blocks, TODO: this could make hole in the data if missing
-    //server is turned off while getting missing blocks
-
-    if (missing > 0) {
-      console.log('missing', missing, 'headers')
-      let missingHeaders = await getPreviousHeaders(
-        missing,
-        lastSavedBlock.number
-      )
+    watcher.ping(enhancedHeader.timestamp, finalizedHash.toString())
```

Comment:
Processing finalized header in new (not finalized) head subscription

Reply:
yes, this is because ping expects finalized hash to check if it changed. we get it  261